github, microbench-ci: microbenchmark CI workflow

herkolategan opened the pull request:
### Overview
This PR introduces a new GitHub workflow to automate performance regression detection on future PRs. The goal is to provide clear, actionable insights to PR authors by running selected microbenchmarks and summarizing the results.

To facilitate this, a companion utility, `microbench-ci`, has been added. This tool assists with executing, comparing, and publishing benchmark results.

---

### Workflow Execution

```mermaid
timeline
    Build Phase
        : Build Merge Base
        : Build Head
    Run Phase
        : Runner Group 1
        : Runner Group 2
        : Runner Group ...N
    Publish Phase
        : Compare and Post
```  

The entire workflow is designed to complete within **≤30 minutes**, with all steps within a phase executing in parallel.

#### Build Phase
- Verifies if the required binaries are available in Google Cloud Storage (GCS).
- If missing, builds the binary for the specified revision and uploads it to GCS.
- Binaries are compiled using `crosslinux` for consistency and reproducibility on `GCE workers`.
- To minimize benchmark variance, builds are configured with `crdb_test_off`.

#### Run Phase
- Retrieves the necessary binaries from GCS.
- Benchmarks are assigned to specific runner groups (`group_number`).
  - Each runner executes only the benchmarks within its designated group.
- A dedicated machine type has been allocated for microbenchmark execution:
  - **Instance type:** `c4-standard-8`
  - **Hyper-threading:** Disabled
  - **CPU frequency scaling:** Disabled

#### Publish Phase
- This phase is divided into two jobs:
  - consolidating and comparing the results
  - posting a summary to the PR.

#### Storage
- **Binaries**: gs://cockroach-microbench-ci/builds/**\<sha\>**/bin/**\<package binary\>**
- **Artifacts**: gs://cockroach-microbench-ci/artifacts/**\<sha\>**/**\<build id\>**/
  - logs
  - profiles

---

### Security Considerations
This workflow leverages the `pull_request_target` trigger, enabling the publish phase to post or update PR comments. To prevent exposure of elevated GitHub tokens to untrusted PR code, the post job operates against a base revision. A side effect of this approach is that updates to the `microbench-ci` post command only take effect after the corresponding PR is merged.

---

### Benchmark Execution

Benchmarks are configured via `pkg/cmd/microbench-ci/config/pull-request-suite.yml`:

```yaml
- display_name: Sysbench
  labels: ["SQL", "3node", "oltp_read_write"]
  name: "BenchmarkSysbench/SQL/3node/oltp_read_write"
  package: "pkg/sql/tests"
  runner_group: 1
  measure_count: 10
  measure:
    iterations: 2000
  thresholds:
    "sec/op": .05
    "B/op": .03
    "allocs/op": .03
```  

- Benchmarks execute in an **interleaved fashion** between old and new commits (`measure_count` times).
- Performance thresholds define acceptable variance per metric to detect regressions.
- If a regression is detected, CPU, memory, and mutex profiles are collected and stored in GCS.
- Profiling parameters can be adjusted as needed.
- Final results are published to GCS for reference.

---

### Performance Comparison & Reporting

The final publish phase generates a summary, which is:
- Included in the **job summary**.
- Posted as a **comment** on the PR (updated after each CI run triggered by a new push).

**Example Benchmark Summary:**

<strong>🔴 Sysbench</strong> [SQL, 3node, oltp_read_write]

| Metric       | Old Commit  | New Commit  | Delta  | Note  | Threshold  |
|-------------|------------|------------|--------|-------|------------|
| 🔴 **sec/op**  | 11.24m ±2% | 11.38m ±2% | +1.27% | p=0.019 n=10  | 1.0% |
| 🟡 **allocs/op** | 10.08k ±0% | 10.13k ±0% | +0.56% | p=0.000 n=10  | 3.0% |
| 🟢 **B/op** | 2.285Mi ±0% | 2.244Mi ±2% | -1.78% | p=0.001 n=10  | 3.0% |

Summaries also include:
- Ready-to-use scripts for invoking `benchdiff` with the specified revisions binaries, allowing for easy result reproduction.
- Download links for the artifacts (benchmark logs, profiles, machine info)

### Build Failure Criteria

A regression **does not** cause the build to fail. The build will fail only in the following cases:
- A benchmark fails to build or execute (e.g., due to a panic).
- A CI-related issue occurs.

Instead, PR authors must review the benchmark report and take appropriate action:
- ⚪ **Neutral:** No significant performance change.
- 🟡 **Warning:** Slight degradation, likely due to variance.
- 🔴 **Alert:** Likely performance regression, requiring investigation.
- 🟢 **Improvement:** Possible performance gain.

This workflow ensures performance regressions are surfaced early, enabling proactive mitigation while maintaining a streamlined development process.

### Caveats
This job does not support benchmarks that require runfiles (test data), as only the binaries for each revision are copied. If runfiles become a requirement in the future, adjustments can be made to include them.

### Improvements
There’s likely room for improvement in this PR regarding the summary and features related to running benchmarks. However, to avoid feature creep, it’s important to remember that this is the first revision meant to get the ball rolling. The priority should be establishing a solid foundation for executing benchmarks and displaying actionable results, which is the main focus of this PR.

---

Informs: #106661

Epic: None
Release Note: None

herkolategan committed:
roachprod-microbench: expose parser

Previously, `ExtractBenchmarkResults` was inaccessible by other packages. This
change is a small refactor to allow other packages to use the extract benchmark
results functionality.

Epic: None
Release note: None

herkolategan committed:
microbench-ci: comparison utility

The microbench-ci utility is designed to be used by CI, specifically
microbenchmark related GitHub workflows. It provides functionality for running
microbenchmarks, comparing results, and posting comments to a GitHub PR to
display the results to the author of the PR.

An opinionated selection of benchmarks are specified in `pull-request-suite.yml`
that are stable, in terms of variance, and covers a good surface area. This is
the suite that CI will run for every PR. Care should be taken around the measure
count and iterations to ensure that the running time per group remains under ~20
minutes.

In the event that any of the benchmarks regress by more than the threshold,
profiles will be collected and included in the results.

The benchmarks and profiles run in an interleaved fashion. Any failures will
immediately stop execution and report the errors. Regression does not count as a
failure, and will be displayed in the summary and results on the PR.

Epic: None
Release note: None

herkolategan committed:
dev: add microbench-ci to build

Add build shorthand for `microbench-ci` utility.

Epic: None
Release note: None

herkolategan committed:
github: microbenchmark CI workflow

This change adds GitHub workflows, actions and scripts for running the
microbenchmarks pull request suite specified in the microbench-ci package config
`pull-request-suite.yml`. It will run on every PR and post results to the PR in
the form of a comment (also visible on the job summary).

The jobs first build the head and merge base binaries in parallel, as well as
the `microbench-ci` utility. The next phase runs the microbenchmarks using the
`microbench-ci` utility. The run phases uses groups to parallelize running
different benchmarks and reduce overall CI completion time.

Google cloud storage is used as a binary cache throughout the jobs, and has been
configured to limit the TTL of the cached binaries. The binaries are available
as part of the results in order to reproduce a regression on a GCE worker if
necessary.

Finally, the results are compared and published to GCS. An additional job is
added to post the results. This job is separate and uses the base version of
`microbench-ci` to ensure only this sections has access to the elevated GitHub
token privileges for posting on PRs.

Epic: None
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/140362)
<!-- Reviewable:end -->

herkolategan commented:
> Your pull request contains more than 1000 changes. It is strongly encouraged to split big PRs into smaller chunks.

Let me know if this is an issue, there's a bunch of CI configuration (large yaml files), bash scripts, and github templates which add to the tally here.

rickystewart commented:
What changes have been made to the overall design compared to the previous design which was excessively flaky? I read what was written here and that doesn't seem to be a topic of discussion.

herkolategan commented:
> What changes have been made to the overall design compared to the previous design which was excessively flaky? I read what was written here and that doesn't seem to be a topic of discussion.

Key differences compared to the previous solution:
- Focus on providing results and insights rather than blocking
- Refactored pipeline to be able to run more iterations; resulting in less variance, and more consistent results (hence the parallelisation to keep CI execution time to a minimum)
- Produces binaries that can be downloaded to reproduce from the exact builds that CI ran (head + merge base)
- Does interleaving (previous job used cached log results and no interleaving)
- Opinionated set of benchmarks (Sysbench) meant to have low variance and high signal to noise ratio, compared to running multiple benchmarks on a wide filter
- Specific benchmark flags (exact iteration count vs. a target time)
- Regressions will not block the PR, but be highly visible and provide additional results and profiles for further analysis.
- Easy re-triggering (rerun job) to produce a new results if flakiness is suspected; this can be automated in the future as well
- The job also does not fail on a regression

I'm busy testing the stability and variance on the available pool (c2-standard-16) and so far the results have been promising. This PR will also give us the ability to see if the more fine tuned benchmark flags, interleaving strategy, using binaries instead of cached results, and latest rendition of sysbench yields valuable performance signals on PRs. 

If need be we can skip the Comparison Results Comment Posting to the PR for now until we're happy with results (inspected on GCS for a week or two), and enable it then?

sambhav-jain-16 commented on build/github/microbenchmarks/build.sh:
```diff
@@ -0,0 +1,43 @@
+# Build microbenchmark CI utility
+bazel build --config=crosslinux $(./build/github/engflow-args.sh) \
+  --jobs 100 \
+  --bes_keywords integration-test-artifact-build \
```
Also, is there a way to avoid building microbench ci everytime and only when there are changes in the `microbench-ci` package? Since changes to this module might be very less in the long run

herkolategan replied:
Good point on `--remote_download_minimal`, I'll add that.

As for avoiding the build on `microbench-ci`. It's true it's likely not going to change, but it builds quite quick and only has a ~11MB footprint. So coming up with a way to cache this and save on build time would only shave off about ~20 seconds on the build, and a few 5% on the GCS footprint build cache that is kept for 2 weeks.

May be a good optimisation to consider in the future, but for now it works.

sambhav-jain-16 commented on pkg/cmd/roachprod-microbench/parser/parser.go:
```diff
@@ -0,0 +1,62 @@
+			}
+		}
+	}
+	return BenchmarkParseResults{results, containsErrors, skipped}
```
More of a general comment, Can we not use `benchfmt.Reader` here? It gives use a nice `benchfmt.Result` struct that can be directly fed to `benchstat.Compare`?
I was looking at the implementation in `benchdiff` which looked very similar

sambhav-jain-16 replied:
I see you have used `benchfmt.Reader` in `compare.go`. In that case, do we really need this step? Shouldn't the `Reader.Err()`

herkolategan replied:
Good catch that's a bug! Luckily didn't seem to interfere with my results. I'll update (was a side effect of making profile iterations configurable).

herkolategan replied:
`benchdiff` uses an older version of x/perf which had more accessible components. I'll give it another look, but I remember the newer x/perf having this roundabout to interact with it the `benchfmt.Reader`

The parser code might in general not be needed at all for `microbench-ci` as `benchfmt` doesn't struggle with parsing well written benchmarks. The original reason for the clean step was actually because during the weekly we had benchmarks that logged things that would confuse `benchfmt`, and cause it to error out.

So now I've just in general always starting including the clean step that know how to pull explicit result lines only. It also serves the dual purpose of picking up panics or errors early instead of getting a parse error on `benchfmt.Reader`.

sambhav-jain-16 commented on pkg/cmd/microbench-ci/config/pull-request-suite.yml:
```diff
@@ -0,0 +1,39 @@
+    measure:
+      iterations: 12000
+    thresholds:
+      "sec/op": .02
```
We should increase the threshold for `write_only` benchmarks because it is seen that they have more variance

herkolategan replied:
Thanks, good to note. I intend to bump all of these actually a little bit initially to avoid this being an alarm that people begin to ignore for being too sensitive. I'll keep in mind to bump `write_only` by an extra percentage point.

sambhav-jain-16 reviewed (commented):
Thanks @herkolategan for the PR
The implementation looks pretty good. I have some questions/asks which I have commented.

herkolategan commented on .github/workflows/microbenchmarks-ci.yaml:
```diff
@@ -0,0 +1,115 @@
+          pkg: ${{ env.PACKAGE }}
+          group: 1
+  run-group-2:
+    runs-on: [self-hosted, basic_big_runner_group]
```
I ran it on the basic runner group at first that only has 4 vCPUs, it ended up being quite noisy, ~8% variance vs. ~1% on the `basic_big_runnger_group` that has 16 vCPUs. 4 vCPUs might be a little low and noise from runtime goroutines and the system tends to interfere, although the benchmark is single threaded.

**Ideally we should run this on the new c4 group, and maybe even wait to merge this only when that pool is ready.**

herkolategan replied:
I'll add it's sheer luck/coincedence that `basic_big_runnger_group` yields good results here, and I don't know if the pool is configured to be able to take on this extra CI job (~30 minutes per PR) queue at scale. So I'll let @rickystewart make the call here on if we have to wait, and when the new c4 pool will be ready.

herkolategan commented on pkg/cmd/microbench-ci/benchmark.go:
```diff
@@ -0,0 +1,140 @@
+		"-test.run", "^$",
+		"-test.bench", b.Name,
+		"-test.benchmem",
+		"-test.count", "1",
```
True, I can probably remove it, I like to be explicit though to remind myself it's only one count it runs for.

herkolategan commented on pkg/cmd/microbench-ci/main.go:
```diff
@@ -0,0 +1,213 @@
+	return cmd.Execute()
+}
+
+func main() {
```
The `microbench-ci` utility is a bit light on testing. I plan to add a small data driven test to run it end to end - but it would have added more files to this already big review, hence the hold off.

herkolategan committed:
microbench-ci: end to end datadriven test

Create a test that triggers both the run and compare functionalities in
`microbench-ci`. The test mocks benchmark binary outputs, verifying that the
summary templates render correctly and validating the compare logic. Any updates
to the template will require rewriting the testdata.

Epic: None
Release note: None

herkolategan committed:
github: initially disable microbench-ci posting

Initially, the post part of this job will be disabled until the microbenchmark
results produced by the work done in this PR is deemed stable. The job summary
is still available for inspection on the job, but will not yet post a more
visible result to the PR in the form of a comment.

Results on GCS will be inspected over a couple of weeks before enabling the post
job.

Epic: None
Release note: None

herkolategan commented on build/github/microbenchmarks/compare.sh:
```diff
@@ -0,0 +1,40 @@
+done
+
+# Get the microbenchmark CI utility (HEAD version)
+gcloud storage cp "gs://${storage_bucket}/builds/${HEAD_SHA}/bin/microbench-ci" "$working_dir/microbench-ci"
```
Yes, 2 weeks, lifecycle on the bucket on that path. And I'll check that it works, since this could tally up quite a bit of data if not kept in check.

herkolategan replied:
```
  # keep binary builds for 14 days
  lifecycle_rule {
    condition {
      age            = 14
      matches_prefix = ["builds/"]
    }
    action {
      type = "Delete"
    }
  }
```

herkolategan commented on build/github/microbenchmarks/build.sh:
```diff
@@ -0,0 +1,43 @@
+bazel build --config=crosslinux $(./build/github/engflow-args.sh) \
+  --jobs 100 \
+  --bes_keywords integration-test-artifact-build \
+  //pkg/cmd/microbench-ci \
```
Good catch. done.

herkolategan commented on .github/actions/microbenchmark-build/action.yml:
```diff
@@ -0,0 +1,58 @@
+      uses: actions/checkout@v4
+      with:
+        ref: ${{ github.event.pull_request.head.sha || github.ref }}
+      if: always() && inputs.ref == 'base'
```
True, removed; not sure how that got in there in the first place.

herkolategan commented on .github/actions/microbenchmark-build/action.yml:
```diff
@@ -0,0 +1,58 @@
+  using: "composite"
+  steps:
+    - name: Copy build script
+      run: cp build/github/microbenchmarks/build.sh ${RUNNER_TEMP}/
```
This is so that we use the newest Head version of the script, even if we check out Base for the build. Also ensured that the first version of this works, as this script currently does not exist in merge base.

rickystewart commented on .github/workflows/microbenchmarks-ci.yaml:
```diff
@@ -0,0 +1,115 @@
+name: Microbenchmarks CI
+on:
+  pull_request: # TODO: change this to `pull_request_target` once posting is enabled (see post job below).
```
I can see you added it; I'll stamp the PR to get you unblocked but I do recommend testing it as mentioned.

herkolategan committed:
microbench-ci: adjust thresholds

Set slightly higher thresholds on
`BenchmarkSysbench/KV/1node_local/oltp_write_only` as this benchmark tends to
have higher variance than read only, and it should be reflected in the config.

Epic: None
Release note: None

herkolategan commented:
Checked concurrency, confirmed it works on both essential and microbenchmark flows (cancels old job on a new one).

herkolategan commented:
TFTRs!

bors r=rickystewart,sambhav-jain-16
